draft-ietf-oauth-v2-1.md

```diff
@@ -616,54 +616,60 @@ for clients that may be used even with manual client registration.
 
 ## Client Types {#client-types}
 
-OAuth 2.1 defines three client types based on their ability to authenticate securely
-with the authorization server as well as the authorization server's assurance of the
-client's identity.
+OAuth 2.1 defines two client types based on their ability to authenticate securely
+with the authorization server.
 
 "confidential":
-: Clients that have credentials and have a prior relationship with the AS are designated as "confidential clients"
-
-"credentialed":
-: Clients that have credentials but no prior relationship with the AS are designated as "credentialed clients"
+: Clients that have credentials with the AS are designated as "confidential clients"
 
 "public":
 : Clients without credentials are called "public clients"
 
 Any clients with credentials MUST take precautions to prevent leakage and abuse of their credentials.
 
+Client authentication allows an Authorization Server to ensure it interacts with a certain client 
```

Comment:
you mean use the secret only in certain scenarios? what would be the reason to do so?

Reply:
Sure, I have a native OAuth client with a security requirement to invalidate ATs when users want to sign out and/or they just close the application. What do I do? I present ATs to the token revocation endpoint which is protected... it isn't desirable just to let anyone else revoke ATs.

Let's take a look at the introspection endpoint too, my native OAuth client needs to get an information associated with an AT such as userId... so, the client needs to be somehow authenticated because again, it's undesirable to let anyone else obtain this kind of information. An alternative approach would be to issue idToken or self-contained ATs and extract everything I need from them. https://datatracker.ietf.org/doc/html/rfc8252 doesn't require to use of the "openid" scope and we don't always deal with self-contained ATs so these approaches won't work all the time.

In a more general sense, a public client may want to use the "client_credentials" grant type which is allowed on OAuth 2.1. So, the client needs to be authenticated for the token endpoint, right? The client might want to get ATs that then could be used for interactions with the abovementioned introspect and revocation endpoints, or with any other resources which could be useful for it.

Right now, the specification defines public clients as "clients without credentials" although, as I tried to show, sometimes it's useful or even required to have and support one or another form of client authentification.

Reply:
Clients shouldn't be using the token introspection endpoint, that endpoint is for RSs to introspect tokens, so it makes sense for that endpoint to require authentication.

As for the revocation endpoint, I feel like this question comes up quite often, the argument being requiring separate authentication reduces the chance of token scanning attacks, but the counterargument being if token scanning attacks are a concern then you have some other design issue with your system and also every RS would be a target for the same scanning attacks, so there really isn't anything new gained by adding separate authentication.

> it isn't desirable just to let anyone else revoke ATs.

Only the bearer of an AT can revoke an AT. That's what makes it a bearer token.

> my native OAuth client needs to get an information associated with an AT such as userId

No, only the RS should be introspecting the access token. The client should get this information from the OpenID Connect ID token.

> In a more general sense, a public client may want to use the "client_credentials" grant type

No, public clients can't use the client credentials grant type because they aren't issued client credentials.

Reply:
Thank you for the response.

[RFC 7662](https://datatracker.ietf.org/doc/html/rfc7662) provides a generic mechanism to validate tokens, including RTs as well. There is a dedicated value for them in the token_type_hint parameter for them. I know that it's a quite popular opinion that the introspection endpoint is for RSs only.

It feels like there is no other way than to issue the Open ID Connect ID token as of now... and what's strange is that I have never seen any discussion around that before... taking into account the popularity of [OAuth 2.0 for Native Apps (RFC 8252)](https://datatracker.ietf.org/doc/html/rfc8252).

Reply:
The same applies to the token revocation process in native apps as well.

Reply:
The first sentence of RFC 7662 is:

> This specification defines a method for a **protected resource** to query
> an OAuth 2.0 authorization server to determine the active state of an
> OAuth 2.0 token and to determine meta-information about this token.

(emphasis mine)

This means it is for resource servers to introspect tokens, not for clients or other parties.

There isn't much discussion of clients obtaining user info in OAuth because the scope of OAuth is authorizing access to APIs. OpenID Connect fills in the parts needed for clients getting user information, so that's where you'll find those things discussed.

Reply:
RFC 7662 gives the following description of the token parameter in the introspection request

```
The string value of the token.  For access tokens, this
      is the "access_token" value returned from the token endpoint
      defined in OAuth 2.0 [RFC6749], Section 5.1.  For refresh tokens,
      this is the "refresh_token" value returned from the token endpoint
      as defined in OAuth 2.0 [RFC6749], Section 5.1.
```

what is the point to deny it? Do you often see RTs at the RS?

Reply:
That's a good point, the RS shouldn't be seeing refresh tokens, so I'm not sure why that's there. But in any case that is an issue with 7662 not related to 2.1 since we're not bringing that in. Can you bring this up on the mailing list?